File: cpp/collatz-conjecture/collatz_conjecture.cpp
#include "collatz_conjecture.h"

#include <iostream>
namespace collatz_conjecture {
  unsigned steps(int n) {
    if (n <= 0) throw std::domain_error("");

    unsigned step = 0;
    while (n ^ 1) {
      n = ~n & 1 ? n >> 1 : n * 3 + 1;
      ++step;
    }
    return step;
  }
}  // namespace collatz_conjecture

File: zig/collatz-conjecture/collatz_conjecture.zig
pub const ComputationError = error{IllegalArgument};

pub fn steps(number: usize) anyerror!usize {
    if (number < 1) {
        return ComputationError.IllegalArgument;
    }
    var n = number;
    var step: usize = 0;
    while (n > 1) : (step += 1) {
        if (n & 1 == 0) {
            n = n / 2;
        } else {
            n = n * 3 + 1;
        }
    }
    return step;
}
